File: 0659-Split Array into Consecutive Subsequences/0659-Split Array into Consecutive Subsequences.cpp
class Solution {
public:
    bool isPossible(vector<int>& nums) {
        unordered_map<int, int> freq, append;
        for (int& num : nums)
        {
            ++freq[num];
        }
        
        for (int& num : nums)
        {
            if (freq[num] == 0)
            {
                continue;
            }
            else if (append[num] > 0)
            {
                --append[num];
                ++append[num + 1];
            }
            else if (freq[num + 1] > 0 && freq[num + 2] > 0)
            {
                --freq[num + 1];
                --freq[num + 2];
                ++append[num + 3];
            }
            else
            {
                return false;
            }
            --freq[num];
        }
        
        return true;
    }
};

File: 0601-0700/0659-Split Array into Consecutive Subsequences/0659-Split Array into Consecutive Subsequences.py
import collections
class Solution:
    def isPossible(self, nums: List[int]) -> bool:
        freq = collections.Counter(nums)
        app = collections.Counter()
        for num in nums:
            if freq[num] == 0:
                continue
            if app[num] > 0:
                app[num] -= 1
                app[num + 1] += 1
            elif freq[num + 1] > 0 and freq[num + 2] > 0:
                freq[num + 1] -= 1
                freq[num + 2] -= 1
                app[num + 3] += 1
            else:
                return False
            freq[num] -= 1
        return True
